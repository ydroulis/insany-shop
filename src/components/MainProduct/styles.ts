"use client"

import styled from 'styled-components'

export const Container = styled.main`
    font-family: Inter, sans-serif;

    padding: 5rem 1rem 10.5rem 1rem;

    @media (min-width: 768px) {
        padding: 5rem 2rem 10.5rem 2rem;

        .wrapper {
            max-width: 704px;
            margin: 0 auto;
        }
    }

    @media (min-width: 1024px) {
        padding: 5rem 3rem 10.5rem 3rem;

        .wrapper {
            max-width: 928px;
            margin: 0 auto;
        }
    }

    @media (min-width: 1440px) {
        padding: 5rem 10rem 10.5rem 10rem;

        .wrapper {
            max-width: 1120px;
            margin: 0 auto;
        }
    }
`

export const Back = styled.button`
    display: flex;
    align-items: center;
    gap: 8px;

    border: none;
    background-color: transparent;
    cursor: pointer;

    font-size: 0.875rem;
    color: #617480;

    margin-top: 1.563rem;
    margin-bottom: 1.375rem;

    &:hover {
        opacity: 0.8;
    }
`

export const Body = styled.section`
   width: 100%;
   display: flex;
   flex-direction: column;
   gap: 32px;

   @media (min-width: 768px) {
        flex-direction: row;
        gap: 16px;
   }

   @media (min-width: 1024px) {
        gap: 32px;
   }
`