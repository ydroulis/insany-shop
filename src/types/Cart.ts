
export type CartProduct = {
    id: number
    name: string
    description: string
    price: number
    image: string
    stock: number
    items?: number
}

export type Cart = {
    content: CartProduct[]
    total: number
}