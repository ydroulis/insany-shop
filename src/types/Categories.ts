export type Categories = Category[];

export type Category = {
    id: string,
    name: string,
    description: string,
    icon: string,
    productCount: number
};