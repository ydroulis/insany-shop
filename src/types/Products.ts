export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}