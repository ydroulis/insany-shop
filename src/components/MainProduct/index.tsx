import React from 'react';

import * as S from './styles';
import { Product } from '@/types/Products';
import ProductImage from '../ProductImage';
import ProductInfo from '../ProductInfo';

import Image from 'next/image';
import { useCategoriesStore } from '../../providers/categoriesStoreProvider';
import { useRouter } from 'next/navigation';

interface MainProductProps {
    product: Product
}

const MainProduct: React.FC<MainProductProps> = ({ product }) => {
    const { categories } = useCategoriesStore((state) => state);
    const router = useRouter();

    const category = categories.find(category => {
        return category.id === product.category;
    })

    const handleBack = () => router.back();

    return (
        <S.Container>
            <S.Back data-testid="back-button" onClick={handleBack} aria-label="Voltar para a página anterior">
                <Image src="/back.png" alt="Ícone voltar" width={24} height={24} />
                Voltar
            </S.Back>
            <S.Body aria-labelledby="product-title" data-testid="body-product">
                <ProductImage image={product.image} name={product.name} />
                <ProductInfo
                    name={product.name}
                    description={product.description}
                    price={product.price}
                    category={category?.name}
                />
            </S.Body>
        </S.Container>
    );
}

export default MainProduct;