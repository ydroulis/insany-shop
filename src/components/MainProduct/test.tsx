import React from "react";
import { render, screen } from "@testing-library/react";
import MainProduct from "./";
import { Product } from "@/types/Products";

jest.mock("next/navigation", () => ({
    useRouter: () => ({
        back: jest.fn(),
    }),
}));

jest.mock("../../providers/categoriesStoreProvider", () => ({
    useCategoriesStore: jest.fn(() => ({
        categories: [
            { id: "1", name: "Casa e Decoração" },
            { id: "2", name: "Eletrônicos" },
        ],
    })),
}));

jest.mock("next/image", () => ({
    __esModule: true,
    default: (props: { src: string; alt: string }) => <img {...props} />,
}));

describe("<MainProduct />", () => {
    const product: Product = {
        id: 1,
        name: "Sofá 3 Lugares Retrátil",
        description:
            "Sofá confortável com assento retrátil e reclinável, revestimento em tecido suede e estrutura de madeira maciça.",
        price: 1899.99,
        image: "/fake-product.jpg",
        category: "1",
        stock: 10,
        rating: 4.5,
    };

    it("should render all main elements", () => {
        render(<MainProduct product={product} />);

        const body = screen.getByTestId("body-product");
        expect(body).toBeInTheDocument();
        expect(body).toHaveAttribute("aria-labelledby", "product-title");

        const backButton = screen.getByRole("button", { name: /voltar para a página anterior/i });
        expect(backButton).toBeInTheDocument();

        const title = screen.getByRole("heading", { name: /Sofá 3 Lugares Retrátil/i });
        expect(title).toBeInTheDocument();

        const image = screen.getByRole("img", { name: /imagem ilustrativa do produto sofá 3 lugares retrátil/i });
        expect(image).toBeInTheDocument();
        expect(image).toHaveAttribute("src", "/fake-product.jpg");

        expect(screen.getByText(/Sofá confortável com assento retrátil/i)).toBeInTheDocument();

        expect(screen.getByText(/Casa e Decoração/i)).toBeInTheDocument();

        const price = screen.getByLabelText(/Preço do produto/i);
        expect(price).toBeInTheDocument();
        expect(price.textContent?.replace(/\u00A0/g, " ")).toBe("R$ 1.899,99");

    });

    it("should match snapshot", () => {
        const { container } = render(<MainProduct product={product} />);
        expect(container).toMatchSnapshot();
    });
});
