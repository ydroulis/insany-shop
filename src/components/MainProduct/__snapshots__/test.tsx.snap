// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<MainProduct /> should match snapshot 1`] = `
<div>
  <main
    class="sc-bRKDuR jUVPqi"
  >
    <button
      aria-label="Voltar para a página anterior"
      class="sc-hvigdm jhQtkF"
      data-testid="back-button"
    >
      <img
        alt="Ícone voltar"
        height="24"
        src="/back.png"
        width="24"
      />
      Voltar
    </button>
    <section
      aria-labelledby="product-title"
      class="sc-fhHczv floSdJ"
      data-testid="body-product"
    >
      <img
        alt="Imagem do produto"
        aria-label="Imagem ilustrativa do produto Sofá 3 Lugares Retrátil"
        blurdataurl="..."
        class="sc-ggWZvA lngDEH"
        height="580"
        loading="lazy"
        placeholder="blur"
        role="img"
        sizes="(max-width: 768px) 100vw, 640px"
        src="/fake-product.jpg"
        width="640"
      />
      <section
        aria-describedby="product-description"
        aria-labelledby="product-title"
        class="sc-dTvVRJ hyRNqX"
      >
        <div
          class="sc-jwTyAe kKULXm"
        >
          <p
            class="sc-hjsuWn cOfegs"
          >
            Casa e Decoração
          </p>
          <h1
            class="sc-jJLAfE bypYEZ"
            id="product-title"
          >
            Sofá 3 Lugares Retrátil
          </h1>
          <p
            aria-label="Preço do produto"
            class="sc-hwkwBN hGDnhX"
          >
            R$ 1.899,99
          </p>
          <div
            class="sc-kNOymR clVJot"
            id="product-description"
          >
            <p>
              Descrição
            </p>
            <span>
              Sofá confortável com assento retrátil e reclinável, revestimento em tecido suede e estrutura de madeira maciça.
            </span>
          </div>
        </div>
        <button
          aria-label="Adicionar"
          class="sc-dYwGCk iBUWmd"
          type="button"
        >
          <svg
            aria-hidden="true"
            color="#fff"
            data-testid="icon-svg"
            fill="none"
            focusable="false"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="color: rgb(255, 255, 255);"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="9"
              cy="21"
              r="1"
            />
            <circle
              cx="20"
              cy="21"
              r="1"
            />
            <path
              d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
            />
          </svg>
          <p>
            Adicionar
          </p>
        </button>
      </section>
    </section>
  </main>
</div>
`;
